cpp/daal/src/algorithms/dtrees/gbt/regression/gbt_regression_predict_dense_default_batch_impl.i

```diff
@@ -207,39 +291,182 @@ protected:
+ * \param[in] nRowsData number of rows to process
+ * \param[in] nColumnsData number of columns in data, i.e. features
+ *                         note: 1 SHAP value per feature + bias term
+ * \param[in] x pointer to the start of observation data
+ * \param[out] res pointer to the start of memory where results are written to
+ * \param[in] dim DimType helper
+*/
+template <typename algorithmFPType, CpuType cpu>
+template <bool hasUnorderedFeatures, bool hasAnyMissing>
+services::Status PredictRegressionTask<algorithmFPType, cpu>::predictContributionInteractions(size_t iTree, size_t nTrees, size_t nRowsData,
+                                                                                              const algorithmFPType * x,
+                                                                                              const algorithmFPType * nominal, algorithmFPType * res,
+                                                                                              const DimType & dim)
+{
+    Status st;
+    const size_t nColumnsData  = dim.nCols;
+    const size_t nColumnsPhi   = nColumnsData + 1;
+    const size_t biasTermIndex = nColumnsPhi - 1;
+
+    const size_t interactionMatrixSize = nColumnsPhi * nColumnsPhi;
+
+    // Allocate buffer for 3 matrices for algorithmFPType of size (nRowsData, nColumnsData)
+    const size_t elementsInMatrix = nRowsData * nColumnsPhi;
+    const size_t bufferSize       = 3 * sizeof(algorithmFPType) * elementsInMatrix;
+    algorithmFPType * buffer      = static_cast<algorithmFPType *>(daal_calloc(bufferSize));
+    if (!buffer)
+    {
+        st.add(ErrorMemoryAllocationFailed);
+        return st;
+    }
+
+    // Get pointers into the buffer for our three matrices
+    algorithmFPType * contribsDiag = buffer + 0 * elementsInMatrix;
+    algorithmFPType * contribsOff  = buffer + 1 * elementsInMatrix;
+    algorithmFPType * contribsOn   = buffer + 2 * elementsInMatrix;
+
+    // Copy nominal values (for bias term) to the condition = 0 buffer
+    PRAGMA_IVDEP
+    PRAGMA_VECTOR_ALWAYS
+    for (size_t i = 0; i < nRowsData; ++i)
+    {
+        contribsDiag[i * nColumnsPhi + biasTermIndex] = nominal[i];
+    }
+
+    predictContributions(iTree, nTrees, nRowsData, x, contribsDiag, 0, 0, dim);
+    for (size_t i = 0; i < nColumnsPhi; ++i)
+    {
+        // initialize/reset the on/off buffers
+        service_memset_seq<algorithmFPType, cpu>(contribsOff, algorithmFPType(0), 2 * elementsInMatrix);
+
+        predictContributions(iTree, nTrees, nRowsData, x, contribsOff, -1, i, dim);
+        predictContributions(iTree, nTrees, nRowsData, x, contribsOn, 1, i, dim);
+
+        for (size_t j = 0; j < nRowsData; ++j)
+        {
+            const unsigned dataRowOffset = j * interactionMatrixSize + i * nColumnsPhi;
+            const unsigned columnOffset  = j * nColumnsPhi;
+            res[dataRowOffset + i]       = 0;
+            for (size_t k = 0; k < nColumnsPhi; ++k)
```

Comment:
Should it really have 3 nested `for` loops?

Reply:
I believe it can be vectorized.

Reply:
Yes. IIRC they have 4 in XGBoost (due to chunking)

Reply:
About vectorization: I'd say story for another day.

Reply:
If the result is symmetric, it should be possible to reduce the inner loop to something like that:
```
res[dataRowOffset + i] = contribsDiag[columnOffset + k];
for (size_t k = i + 1; k < nColumnsPhi; ++k) {
   auto val = (contribsOn[columnOffset + k] - contribsOff[columnOffset + k]) / 2.0f;
   res[dataRowOffset + k] = val;
   res[j * interactionMatrixSize + k * nColumnsPhi + i] = val;
   res[dataRowOffset + i] -= 2 * val;
}
```